Occassional Unhandled CancellationException in LSPTextHover

When hovering over different parts of the document the following exception is caught:
```
ERROR 2025-11-26 21:00:00.313 [main] org.eclipse.equinox.logger - java.util.concurrent.CancellationException
java.util.concurrent.CompletionException: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:636)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2512)
	at org.eclipse.lsp4e.operations.hover.LSPTextHover.cancel(LSPTextHover.java:213)
	at org.eclipse.lsp4e.operations.hover.LSPTextHover.initiateHoverRequest(LSPTextHover.java:231)
	at org.eclipse.lsp4e.operations.hover.LSPTextHover.getHoverRegion(LSPTextHover.java:152)
	at org.eclipse.ui.internal.genericeditor.hover.CompositeTextHover.getHoverRegion(CompositeTextHover.java:101)
	at org.eclipse.jface.text.TextViewerHoverManager.computeInformation(TextViewerHoverManager.java:128)
	at org.eclipse.jface.text.AbstractInformationControlManager.doShowInformation(AbstractInformationControlManager.java:1101)
	at org.eclipse.jface.text.AbstractHoverInformationControlManager$MouseTracker.mouseHover(AbstractHoverInformationControlManager.java:511)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:246)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:91)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4362)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1214)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4160)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3748)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1042)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:678)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:583)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	[...]
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:219)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:149)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:115)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:467)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:298)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:627)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:575)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1431)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1403)
Caused by: java.util.concurrent.CancellationException: null
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2510)
	... 33 common frames omitted
```
I think we need a catch clause for CancellationException here, where we return null?
https://github.com/eclipse-lsp4e/lsp4e/blob/70e805edeb01edc367ae49921a2b3c435c396146/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java#L173-L179

--- a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/FocusableBrowserInformationControl.java
+++ b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/FocusableBrowserInformationControl.java
@@ -26,6 +26,7 @@ import org.eclipse.jface.text.IInformationControlCreator;
 import org.eclipse.jface.util.Util;
 import org.eclipse.lsp4e.LSPEclipseUtils;
 import org.eclipse.lsp4e.LanguageServerPlugin;
+import org.eclipse.lsp4e.internal.CancellationUtil;
 import org.eclipse.lsp4e.ui.UI;
 import org.eclipse.swt.browser.Browser;
 import org.eclipse.swt.browser.LocationEvent;
@@ -171,6 +172,11 @@ public class FocusableBrowserInformationControl extends BrowserInformationContro
 					return; // input changed; ignore stale update
 				}
 				if (ex != null) {
+					if (CancellationUtil.isRequestCancelledException(ex)) {
+						// Hover request was cancelled; treat as benign and hide placeholder.
+						super.setInput(""); //$NON-NLS-1$
+						return;
+					}
 					LanguageServerPlugin.logError(ex);
 					super.setInput(
 							"Unexpected error: " + ex.getClass().getSimpleName() + ": " + ex.getLocalizedMessage()); //$NON-NLS-1$ //$NON-NLS-2$

--- a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java
+++ b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/hover/LSPTextHover.java
@@ -21,6 +21,7 @@ import static org.eclipse.lsp4e.internal.NullSafetyHelper.castNonNull;
 import java.util.List;
 import java.util.NoSuchElementException;
 import java.util.Objects;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
@@ -48,6 +49,7 @@ import org.eclipse.jface.text.Region;
 import org.eclipse.lsp4e.LSPEclipseUtils;
 import org.eclipse.lsp4e.LanguageServerPlugin;
 import org.eclipse.lsp4e.LanguageServers;
+import org.eclipse.lsp4e.internal.CancellationUtil;
 import org.eclipse.lsp4j.Hover;
 import org.eclipse.lsp4j.HoverParams;
 import org.eclipse.lsp4j.MarkedString;
@@ -78,8 +80,12 @@ public class LSPTextHover implements ITextHover, ITextHoverExtension, ITextHover
 		if (hoverInfoRequest_.isDone()) {
 			try {
 				return hoverInfoRequest_.getNow(null);
-			} catch (Exception e) {
-				LanguageServerPlugin.logError(e);
+			} catch (final Exception ex) {
+				if (CancellationUtil.isRequestCancelledException(ex)) {
+					// Hover was cancelled; ignore as this is an expected fast-mouse-move scenario.
+					return null;
+				}
+				LanguageServerPlugin.logError(ex);
 			}
 		}
 		return null;
@@ -171,10 +177,12 @@ public class LSPTextHover implements ITextHover, ITextHoverExtension, ITextHover
 					LSPEclipseUtils.toOffset(range.getEnd(), document));
 			return this.lastRegion = new Region(regionStartOffset, regionEndOffset - regionStartOffset);
 		} catch (ExecutionException | BadLocationException e) {
-			LanguageServerPlugin.logError("Cannot get hover region for offset " + offset, e); //$NON-NLS-1$
+			if (!CancellationUtil.isRequestCancelledException(e)) {
+				LanguageServerPlugin.logError("Cannot get hover region for offset " + offset, e); //$NON-NLS-1$
+			}
 		} catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
-		} catch (NoSuchElementException | TimeoutException e) {
+		} catch (NoSuchElementException | TimeoutException | CancellationException e) {
 			// Fallback to heuristic region without blocking.
 		}
 

